### ISSUE TITLE
Global grpc initialization issue

### ISSUE THREAD

**Issue by judywangsg** [grpc/grpc#25020]
### What version of gRPC and what language are you using?
grpc 1.21 c++

### What operating system (Linux, Windows,...) and version?
Oracle Linux Server (version 7.5)

### What runtime / compiler are you using (e.g. python version or version of gcc)
gcc 8.2.0

### What did you do?
We're using grpc to communicate between two parties (java on server and c++ on client). On the client application, compiled **static grpc libraries**(.a files) are used.

### What did you expect to see?
The application can be always running without any crash

### What did you see instead?
The application crashes when it is running at very first time after being installed, after auto recovery it's fine.
From the core file, the crash happens during grpc::CreateCustomChannel(...)
Attached the source code snippt:
```
    ChannelArguments channelArgs;
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, KEEPALIVE_TIME_MS);
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_TIMEOUT_MS, KEEPALIVE_TIMEOUT_MS);
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS, 1);
    channelArgs.SetInt(GRPC_ARG_SERVER_HANDSHAKE_TIMEOUT_MS, HANDSHAKE_TIMEOUT_MS);
    channelArgs.SetCompressionAlgorithm(GRPC_COMPRESS_GZIP);
    std::shared_ptr<grpc_impl::ChannelCredentials> channelCreds  = grpc::InsecureChannelCredentials();
    auto channel = grpc::CreateCustomChannel(address, channelCreds, channelArgs);  => crash here
```
After checking the grpc source code, it seems to be related to global grpc_init(), as the causing of the crash is due to g_state is null
```
 UniquePtr<char> ResolverRegistry::AddDefaultPrefixIfNeeded(const char* target) 
{ 
    GPR_ASSERT(g_state != nullptr);  // =>crash due to g_state is null
    grpc_uri* uri = nullptr;
    char* canonical_target = nullptr;
    g_state->FindResolverFactory(target, &uri, &canonical_target);
    grpc_uri_destroy(uri);
    return UniquePtr<char>(canonical_target == nullptr ? gpr_strdup(target): canonical_target);
}
```
while g_state is initilized by InitRegistry() which is invoked by global **grpc_init()**
```
static RegistryState* g_state = nullptr;
void ResolverRegistry::Builder::InitRegistry()
{ 
   if (g_state == nullptr) g_state = New<RegistryState>(); 
}
```
 So seems the global grpc library initilization is not automatically triggered
I tried to explicitly invok grpc_init() before calling any grpc method and grpc_shutdown() after all grpc work done, it ended up with another crash at grpc_call_start_batch() duringthe first running
```
#1  <signal handler called>
#2  0x00007fffb7f55126 in ?? ()
   from /opt/tbricks/persistence/140742b0-2dff-11eb-87d0-27b7fd406ab0/plugins/2020-12-17-04-23-02-239189/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so
#3  0x00007fffbb28ecce in ?? ()
   from /opt/tbricks/persistence/140742b0-2dff-11eb-87d0-27b7fd406ab0/plugins/2020-12-17-00-00-07-956398/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so
#4  0x00007fffbb28ef37 in grpc_core::ExecCtx::Flush() ()
   from /opt/tbricks/persistence/140742b0-2dff-11eb-87d0-27b7fd406ab0/plugins/2020-12-17-00-00-07-956398/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so
#5  0x00007fffbb2a22c5 in grpc_call_start_batch ()
   from /opt/tbricks/persistence/140742b0-2dff-11eb-87d0-27b7fd406ab0/plugins/2020-12-17-00-00-07-956398/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so
#6  0x00007fffb7c52051 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFillOpsAfterInterception (this=0x7fffb59200c0)
    at /opt/tbricks/apps/third_party/include/grpcpp/impl/codegen/call_op_set.h:940
#7  0x00007fffb7c4f1c3 in PerformOps (this=0x7fffbe669530,
    ops=0x7fffb81ce720 <vtable for grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallNoOp<2>, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >+16>)
    at /opt/tbricks/apps/third_party/include/grpcpp/impl/codegen/call.h:69
#8  ClientReaderWriter (this=<optimized out>, channel=<optimized out>, method=...,
    context=0x7fffb5920370)
    at /opt/tbricks/apps/third_party/include/grpcpp/impl/codegen/sync_stream.h:562
#9  Create (channel=0x7fffbe669550, method=..., context=0x7fffb5920370)
    at /opt/tbricks/apps/third_party/include/grpcpp/impl/codegen/sync_stream.h:435
```

### Anything else we should know about your project / environment?
Our client application is an executable plug-in running inside a complex engine , it works as a shared library which can be dynamically loaded into the engine to be executed. 
Our client application doesn't have its own process, the process is owned by the engine

**Comment by Five9Voice** [comment]
Experiencing a similar segfault.

Code:
channel = grpc::CreateChannel("localhost", grpc::InsecureChannelCredentials());

Backtrace:
(gdb) bt
#0  __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:50
#1  0x00007f067ca82535 in __GI_abort () at abort.c:79
#2  0x00007f06226f5726 in grpc_core::ResolverRegistry::AddDefaultPrefixIfNeeded(char const*) () from /usr/local/mod_grpc.so
#3  0x00007f0622719d6f in grpc_core::(anonymous namespace)::CreateChannel(char const*, grpc_channel_args const*) () from /usr/local/mod_grpc.so
#4  0x00007f0622719fa9 in grpc_insecure_channel_create () from /usr/local/mod_grpc.so

**Comment by Five9Voice** [comment]
Hacked workaround:

Code:

grpc_init();
channel = grpc::CreateChannel("localhost:1234", grpc::InsecureChannelCredentials());

**Comment by yashykt** [comment]
Hi,

I am not able to reproduce this. (Note that I'm using the current gRPC head for testing.) Please update your version if needed. 

Calling `grpc_init` can be a temporary workaround but that should not be a permanent solution. The right thing to do would be to figure out why grpc_init is not being called.

**Comment by jairaju** [comment]
I am seeing similar to what @judywangsg is seeing with shared object library in a client process

env:
grpc 1.28.1 c++
gcc 8.3.1

I am trying to integrate the client code with Dialgoflow CX v3beta1
my code flow is like this

create resources required --
     context = new grpc::ClientContext();
     cq = new grpc::CompletionQueue();
     request = new StreamingDetectIntentRequest();
     response = new StreamingDetectIntentResponse();

     get token -------------

     auto creds = grpc::AccessTokenCredentials(token);
     auto channel_creds = grpc::SslCredentials(grpc::SslCredentialsOptions());

     if (creds.get()) {
           auto cc = grpc::CompositeChannelCredentials(channel_creds, creds);
           auto channel = grpc::CreateChannel("us-central1-dialogflow.googleapis.com", cc);
      }

and with that code i see that grpc_init() is getting called before aborting with this back trace:

#0  0x00007f8ee897a3d7 in raise () from /lib64/libc.so.6
#1  0x00007f8ee897bac8 in abort () from /lib64/libc.so.6
#2  0x00007f8e820cd06b in grpc_core::ResolverRegistry::AddDefaultPrefixIfNeeded (target=<optimized out>) at src/core/ext/filters/client_channel/resolver_registry.cc:185
#3  0x00007f8e820db38a in CreateChannel (args=0x7f8ec0076860, target=0x7f8ec0149dd8 "us-central1-dialogflow.googleapis.com")
    at src/core/ext/transport/chttp2/client/secure/secure_channel_create.cc:161
#4  grpc_secure_channel_create (creds=<optimized out>, target=0x7f8ec0149dd8 "us-central1-dialogflow.googleapis.com", args=0x7f8e7b0bff00, reserved=<optimized out>)
    at src/core/ext/transport/chttp2/client/secure/secure_channel_create.cc:217
#5  0x00007f8e82061f5f in grpc_impl::SecureChannelCredentials::CreateChannelWithInterceptors(std::string const&, grpc_impl::ChannelArguments const&, std::vector<std::unique_ptr<grpc::experimental::ClientInterceptorFactoryInterface, std::default_delete<grpc::experimental::ClientInterceptorFactoryInterface> >, std::allocator<std::unique_ptr<grpc::experimental::ClientInterceptorFactoryInterface, std::default_delete<grpc::experimental::ClientInterceptorFactoryInterface> > > >) () at src/cpp/client/secure_credentials.cc:65
#6  0x00007f8e82062070 in grpc_impl::SecureChannelCredentials::CreateChannelImpl(std::string const&, grpc_impl::ChannelArguments const&) () at src/cpp/client/secure_credentials.cc:51
#7  0x00007f8e82066381 in grpc_impl::CreateCustomChannelImpl(std::string const&, std::shared_ptr<grpc_impl::ChannelCredentials> const&, grpc_impl::ChannelArguments const&) ()
    at src/cpp/client/create_channel.cc:42
#8  0x00007f8e820664df in grpc_impl::CreateChannelImpl(std::string const&, std::shared_ptr<grpc_impl::ChannelCredentials> const&) () at src/cpp/client/create_channel.cc:33
#9  0x00007f8e81f72ff9 in grpc::CreateChannel (target=..., creds=...) at /home/jjayaraju/xs.c49fd3.rh7/grpc-1.28.1.831065/include/grpcpp/create_channel.h:30


Please let me know if there is any progress on the understanding and the solution for it

**Comment by yashykt** [comment]
@jairaju Please verify whether the issue persists at head

**Comment by jairaju** [comment]
@yashykt . I am unable to compile c++ code with the head as it asks for Bazel build which i have not done so far.
I wanted to share our environment with bit more details

In our paltform we support multiple client engines that interact with google servers(ex: dialogflow CX)
each such engine is loaded with a shared object that links grpc code individually but they all run under the same process and are loaded using dlopen().


question:
In this setup,  do i need to do grpc_init() in every such engin? what is suggested here? is there any problem if i do it mutliple times?

actually i tried both doing individual grpc_init() once in each shared object and doing it once per process.
both end up in the same result with a segmentation fault as follows:

#0  push_last_on_exec_ctx (lock=0x7f4a7814de40) at ./src/core/lib/iomgr/exec_ctx.h:145
#1  combiner_exec (error=0x0, cl=0x7f4a78036500, lock=0x7f4a7814de40) at src/core/lib/iomgr/combiner.cc:143
#2  grpc_core::Combiner::Run (this=0x7f4a7814de40, closure=0x7f4a78036500, error=0x0) at src/core/lib/iomgr/combiner.cc:333
#3  0x00007f4a25d9a438 in CheckConnectivityState (try_to_connect=true, this=<optimized out>) at src/core/ext/filters/client_channel/client_channel.cc:4000
#4  grpc_core::(anonymous namespace)::CallData::PickSubchannelLocked (this=this@entry=0x7f4a78171720, elem=elem@entry=0x7f4a78171700,
    error=error@entry=0x7f4a1d587b28) at src/core/ext/filters/client_channel/client_channel.cc:3915
#5  0x00007f4a25d9ad04 in grpc_core::(anonymous namespace)::CallData::PickSubchannel (arg=0x7f4a78171700, error=<optimized out>)
    at src/core/ext/filters/client_channel/client_channel.cc:3898
#6  0x00007f4a285f7631 in exec_ctx_run (closure=<optimized out>, closure=<optimized out>, error=0x0) at src/core/lib/iomgr/exec_ctx.cc:153
#7  grpc_core::ExecCtx::Flush (this=this@entry=0x7f4a1d587bb0) at src/core/lib/iomgr/exec_ctx.cc:153
#8  0x00007f4a2860f875 in ~ExecCtx (this=0x7f4a1d587bb0, __in_chrg=<optimized out>) at ./src/core/lib/iomgr/exec_ctx.h:124
#9  grpc_call_start_batch (call=0x7f4a78170da0, ops=0x7f4a1d587c60, nops=1, tag=0x7f4a78171ee8, reserved=<optimized out>) at src/core/lib/surface/call.cc:1973
#10 0x00007f4a25c4ee40 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpClientSendClose, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFillOpsAfterInterception (this=0x7f4a78171ee8)
    at /home/jjayaraju/xs.c49fd3.rh7/grpc-1.28.1.831065/include/grpcpp/impl/codegen/call_op_set.h:938
#11 0x00007f4a25c4ec02 in grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallOpClientSendClose, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FillOps (this=0x7f4a78171ee8, call=0x7f4a78171c00)
    at /home/jjayaraju/xs.c49fd3.rh7/grpc-1.28.1.831065/include/grpcpp/impl/codegen/call_op_set.h:868
#12 0x00007f4a25c4650f in grpc::internal::Call::PerformOps (this=0x7f4a78171c00, ops=0x7f4a78171ee8)
    at /home/jjayaraju/xs.c49fd3.rh7/grpc-1.28.1.831065/include/grpcpp/impl/codegen/call.h:69
#13 0x00007f4a25c502fd in grpc_impl::ClientAsyncReaderWriter<google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentRequest, google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentResponse>::StartCallInternal (this=0x7f4a78171be0, tag=0x7f4a780494b0)
    at /home/jjayaraju/xs.c49fd3.rh7/grpc-1.28.1.831065/include/grpcpp/impl/codegen/async_stream_impl.h:621
#14 0x00007f4a25e7fff3 in grpc_impl::ClientAsyncReaderWriter<google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentRequest, google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentResponse>::ClientAsyncReaderWriter(grpc::internal::Call, grpc_impl::ClientContext*, bool, void*) () from /home/holly/lib/sr_dfcx.so
#15 0x00007f4a25e7c209 in grpc_impl::internal::ClientAsyncReaderWriterFactory<google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentRequest, google::cloud::dialogflow::cx::v3beta1::StreamingDetectIntentResponse>::Create(grpc::ChannelInterface*, grpc_impl::CompletionQueue*, grpc::internal::RpcMethod const&, grpc_impl::ClientContext*, bool, void*) () from /home/holly/lib/sr_dfcx.so
#16 0x00007f4a25e786b9 in google::cloud::dialogflow::cx::v3beta1::Sessions::Stub::AsyncStreamingDetectIntentRaw(grpc_impl::ClientContext*, grpc_impl::CompletionQueue*, void*) () from /home/holly/lib/sr_dfcx.so
#17 0x00007f4a25c4affc in google::cloud::dialogflow::cx::v3beta1::Sessions::Stub::AsyncStreamingDetectIntent (this=0x7f4a7806dfc0, context=0x7f4a78082380,
    cq=0x7f4a78043af0, tag=0x7f4a780494b0) at /home/jjayaraju/xs.c49fd3.rh7/googleapis-1.0.1.19b4f2/gens/google/cloud/dialogflow/cx/v3beta1/session.grpc.pb.h:208





-------------------------------
one thing to note here is the shared object that is loaded first does not have this problem. the shared object that are loaded later has this problem.
in other words,  if i make use of the engine that is loaded first does not show this behavior, but only the later ones.

Please advice.

**Comment by Frankie0701** [comment]
Hi @yashykt ,

I am team member of @judywangsg , we recently upgraded to v1.36.x, the issue still happens.

`#1  <signal handler called>`
`#2  0x00007ffff093b3d7 in raise () from /usr/lib64/libc.so.6`
`#3  0x00007ffff093cac8 in abort () from /usr/lib64/libc.so.6`
`#4  0x00007fffb958e385 in grpc_core::ResolverRegistry::AddDefaultPrefixIfNeeded(char const*) () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`
`#5  0x00007fffb95a9f96 in grpc_core::(anonymous namespace)::CreateChannel(char const*, grpc_channel_args const*, grpc_error**) () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`
`#6  0x00007fffb95aa1f5 in grpc_insecure_channel_create () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`
`#7  0x00007fffb96b705b in grpc::(anonymous namespace)::InsecureChannelCredentialsImpl::CreateChannelWithInterceptors(std::string const&, grpc::ChannelArguments const&, std::vector<std::unique_ptr<grpc::experimental::ClientInterceptorFactoryInterface, std::default_delete<grpc::experimental::ClientInterceptorFactoryInterface> >, std::allocator<std::unique_ptr<grpc::experimental::ClientInterceptorFactoryInterface, std::default_delete<grpc::experimental::ClientInterceptorFactoryInterface> > > >) () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`
`#8  0x00007fffb96b6fb6 in grpc::(anonymous namespace)::InsecureChannelCredentialsImpl::CreateChannelImpl(std::string const&, grpc::ChannelArguments const&) () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`
`#9  0x00007fffb96b5d51 in grpc::CreateCustomChannel(std::string const&, std::shared_ptr<grpc::ChannelCredentials> const&, grpc::ChannelArguments const&) () from /opt/tbricks/persistence/a6cda586-2dfe-11eb-87d0-27b7fd406ab0/plugins/2021-03-11-02-30-46-166046/2b0537b8-8c13-11e9-9a20-b747deb3124d/DashboardMonitor.so`

**Comment by yashykt** [comment]
I'll do a bit of an explanation on how we expect `grpc_init` to be called normally.

gRPC C++'s classes like `grpc::ChannelCredentials` derive from `grpc::GrpcLibraryCodegen`
https://github.com/grpc/grpc/blob/0937cb1249129a84159c639c59f49a451157edf7/include/grpcpp/security/credentials.h#L69

And within the constructor of `grpc::GrpcLibraryCodegen` we invoke `g_glip->init()` https://github.com/grpc/grpc/blob/0937cb1249129a84159c639c59f49a451157edf7/include/grpcpp/impl/codegen/grpc_library.h#L46

which finally invokes `grpc_init()` https://github.com/grpc/grpc/blob/0937cb1249129a84159c639c59f49a451157edf7/include/grpcpp/impl/grpc_library.h#L34

> In our paltform we support multiple client engines that interact with google servers(ex: dialogflow CX)
each such engine is loaded with a shared object that links grpc code individually but they all run under the same process and are loaded using dlopen().

I suspect loading gRPC in this manner is causing issues but I am not familiar with this enough to be able to pinpoint the issue here.

> On the client application, compiled static grpc libraries(.a files) are used.

@Frankie0701 @judywangsg @Five9Voice Can you please also provide more details as to how gRPC is being built/used?

@veblush @nicolasnoble If y'all have any ideas here.

**Comment by jairaju** [comment]
@yashykt the document here says grpc_init() should be called before calling any other functions, so i am calling it in every shared object in the process.

please see below

https://grpc.github.io/grpc/core/grpc_8h.html#a573aed8e200997856a227c7eb798a185

====snippet==========
Initialize the grpc library.

After it's called, a matching invocation to grpc_shutdown() is expected.

It is not safe to call any other grpc functions before calling this. (To avoid overhead, little checking is done, and some things may work. We do not warrant that they will continue to do so in future revisions of this library).
==================

I understand that even if it is called multiple times, it does not actualy initialize mutliple times as per the code.

**Comment by Frankie0701** [comment]
@yashykt I followed BUILDING.md, and choose cmake to build grpc v1.36.
I changed the **grpc/CMakeList.txt** to add these flags, as it's mandatory for our application to run in the engine.
`set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} ${_gRPC_C_CXX_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC")`
`set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${_gRPC_C_CXX_FLAGS} -D_GLIBCXX_USE_CXX11_ABI=0 -fPIC")`
and it's build with **-DCMAKE_CXX_STANDARD=17**. Just these 2 difference from the grpc installation guide.
After the upgrade, didn't change how grpc is used, it's same as what @judywangsg has posted, just one change on namespace for ChannelCredentials

> ChannelArguments channelArgs;
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_TIME_MS, KEEPALIVE_TIME_MS);
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_TIMEOUT_MS, KEEPALIVE_TIMEOUT_MS);
    channelArgs.SetInt(GRPC_ARG_KEEPALIVE_PERMIT_WITHOUT_CALLS, 1);
    channelArgs.SetInt(GRPC_ARG_SERVER_HANDSHAKE_TIMEOUT_MS, HANDSHAKE_TIMEOUT_MS);
    channelArgs.SetCompressionAlgorithm(GRPC_COMPRESS_GZIP);
    std::shared_ptr<grpc::ChannelCredentials> channelCreds  = grpc::InsecureChannelCredentials();
    auto channel = grpc::CreateCustomChannel(address, channelCreds, channelArgs);  => crash here

**Comment by yashykt** [comment]
> @yashykt the document here says grpc_init() should be called before calling any other functions, so i am calling it in every shared object in the process once in the beginning after it loads the shared objects

My understanding is that you are using C++, and if so, you shouldn't need to call grpc_init()/grpc_shutdown(). The C++ classes are responsible for calling grpc_init/grpc_shutdown.

**Comment by jairaju** [comment]
I have some news about this, If it benefits anybody, it is worth to mention here. 
I was looking through other tickets similar to #21471 and found that when somebody has my kind of environment with multiple shared object under the same process, it is good to keep a single copy of all the grpc libraries.

I build each individual library such as grpcpp_channelz, grpc++, grpc and gpr that were required for the clinet integration in my case to be shared object( earlier i was linking this as static library) and then dynamically link these to all the sub-modules that generate bigger shared object with other client code.

Like @yashykt mentioned i removed the grpc_init() calls in all the sub-modules as it is done internally with other c++ classes such as 
grpc::SslCredentials(grpc::SslCredentialsOptions()); 

but in my case grpc_init() is called in 
cq = new grpc::CompletionQueue(); 
and it seems to be working alright( i don't see those grpc_init() related segmentation faults anymore), although it is initial stages of testing at my end, but it looks like technically this approach is good to maintain one copy of grpc libraries under one process, let the sub-modules share the same copy using shared objects.

I will update if the test results change when more testing is done.
@yashykt let me know if you have any opinions/concerns on this approach

**Comment by yashykt** [comment]
@jairaju Glad to see you got it working :)

**Comment by yashykt** [comment]
Is this still an issue?
